lib/linked_list.rb

```diff
@@ -7,39 +7,76 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    return @head if @head.nil?
+    
+    return @head.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n) where n is the length of the linked list
+  # Space complexity - O(1)
   def length
-    return 0
+    nodes_count = 0
+    current = @head
+
+    until current.nil?
+      nodes_count += 1
+      current = current.next
+    end
+
+    return nodes_count
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n) where n is the length of the linked list
+  # Space complexity - O(1)
   def add_last(data)
+    current = @head
+
+    if @head.nil?
+      add_first(data)
+      return
+    end
 
+    (length - 1).times do 
```

Comment:
This does add an extra traversal to the method 2n instead of n.  It doesn't change the Big-O, but it's a bit inefficient.  Can you think of a better way?